infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/LogReplicationSinkManager.java

```diff
@@ -0,0 +1,286 @@
+    }
+
+    @Override
+    public LogReplicationEntry receive(LogReplicationEntry message) {
+
+        //If the roletype is active or it is not the leader, skip processing the message and ack the timestamp.
+        //This will  trigger the active site to redo the query leadership.
+        if (active || !leader) {
+            LogReplicationEntryMetadata metadata = sinkBufferManager.makeAckMessage(message);
+            return new LogReplicationEntry(metadata, new byte[0]);
+        }
+
+        log.info("Sink manager received {} while in {}", message.getMetadata().getMessageMetadataType(), rxState);
+
+        siteConfigID = Math.max(siteConfigID, message.getMetadata().getSiteConfigID());
+
+        if (!receivedValidMessage(message)) {
+            // If we received a start marker for snapshot sync while in LOG_ENTRY_SYNC, switch rxState
+            if (message.getMetadata().getMessageMetadataType().equals(MessageType.SNAPSHOT_START)) {
+                rxState = RxState.SNAPSHOT_SYNC;
+            } else {
+                // Invalid message // Drop the message
+                log.warn("Sink Manager in state {} and received message {}. Dropping Message.", rxState,
+                        message.getMetadata());
+
+                return null;
+            }
+        }
+
+        if (rxState.equals(RxState.LOG_ENTRY_SYNC)) {
+            return sinkBufferManager.processMsgAndBuffer(message);
+        }
+
+        // Snapshot Sync (no buffering, out of order is allowed)
+        return receiveWithoutBuffering(message);
+    }
+
+    @Override
+    public LogReplicationEntry receive(List<LogReplicationEntry> messages) {
+        for (LogReplicationEntry msg : messages) {
+            receive(msg);
+        }
+
+        // Todo (hack): ACK should be controlled from the receive method
+        if (messages.size() > 0) {
+            return LogReplicationEntry.generateAck(messages.get(0).getMetadata());
+        }
+
+        return null;
+    }
+
+    /**
+     * Signal the manager a snapshot sync is about to complete. This is required to transition to log sync.
+     */
+    private LogReplicationEntry completeSnapshotApply(LogReplicationEntry inputEntry) {
+        log.debug("Complete of a snapshot apply");
+        //check if the all the expected message has received
+        rxState = RxState.LOG_ENTRY_SYNC;
+        persistedWriterMetadata.setSrcBaseSnapshotDone(inputEntry);
+
+        // Prepare and Send Snapshot Sync ACK
+        LogReplicationEntryMetadata metadata = new LogReplicationEntryMetadata(MessageType.SNAPSHOT_END,
+                siteConfigID,
+                snapshotRequestId,
+                persistedWriterMetadata.getLastProcessedLogTimestamp(),
+                persistedWriterMetadata.getLastSrcBaseSnapshotTimestamp(),
+                snapshotRequestId);
+
+        sinkBufferManager = new SinkBufferManager(LOG_ENTRY_MESSAGE, ackCycleTime, ackCycleCnt, bufferSize,
```

Comment:
Do we really need a new instance or is it a reset which sets the new base snapshot?